src/Featurizers/LagLeadOperatorFeaturizer.h

```diff
@@ -0,0 +1,235 @@
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    LagLeadOperatorTransformer(std::uint32_t horizon, std::vector<std::int64_t> offsets);
+    LagLeadOperatorTransformer(Archive &ar);
+
+    ~LagLeadOperatorTransformer(void) override = default;
+
+    FEATURIZER_MOVE_CONSTRUCTOR_ONLY(LagLeadOperatorTransformer);
+
+    bool operator==(LagLeadOperatorTransformer const &other) const;
+    bool operator!=(LagLeadOperatorTransformer const &other) const;
+
+    void save(Archive &ar) const override;
+
+private:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Types
+    // |
+    // ----------------------------------------------------------------------
+    using GrainType  = std::vector<std::string>;
+    // TODO: change target type to general types
+    using TargetType = std::double_t;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Data
+    // |
+    // ----------------------------------------------------------------------
+    Components::CircularBuffer<TargetType>  _buffer;
+    std::int64_t const                      _max_future_offset;
+    std::int64_t const                      _max_prior_offset;
+    std::uint32_t const                     _horizon;
+    std::vector<std::int64_t> const         _offsets;
+    std::int64_t                            _index;
+    std::uint32_t                           _numPending;
+    
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Methods
+    // |
+    // ----------------------------------------------------------------------
+
+    // MSVC runs into problems when the declaration and definition are separated
+    void execute_impl(typename BaseType::InputType const &input, typename BaseType::CallbackFunction const &callback) override {
+        if (_max_future_offset != 0 && _index <= _max_future_offset + _horizon - 1) {
+            ++_index;
+            ++_numPending;
+            _buffer.push(std::get<1>(input));
+            return;
+        }
+        Microsoft::Featurizer::RowMajMatrix<std::double_t> ret(_offsets.size(), _horizon);
+        for (std::uint32_t col = 0; col < _horizon; ++col) {
+            for (size_t row = 0; row < _offsets.size(); ++row) {
+                if ((_index - _numPending - (_horizon - 1 - col) + _offsets[row] < 0)) {
+                    ret(static_cast<long long>(row), col) = std::nan("");
+                }
+                else {
+                    ret(static_cast<long long>(row), col) = *std::get<0>(_buffer.range(
+                                                                                        1, 
+                                                                                        static_cast<size_t>(static_cast<long long>(_buffer.size() - _numPending - (_horizon - 1 - col)) + _offsets[row])
```

Comment:
>_numPending - (_horizon - 1 - col) [](start = 148, length = 34)

Hard to read. So if horizon is 1 and col is 1 then the result is -1 which then add to the result, is it what you  want? Is it possible for the result to be negative. Bc if if you cast negative to size_t the result will be a very big positive integer.

Reply:
Col is also governed by horizon so (_horizon - 1 - col) will always be a positive number

Reply:
And this is a very good point. I just changed it to make it more readable.